batnode.js

```diff
@@ -191,7 +190,103 @@ class BatNode {
       })
     })
   }
-  
+
+  auditFile(manifestFilePath) {
+    const manifest = fileUtils.loadManifest(manifestFilePath);
+    const shards = manifest.chunks;
+    const shaIds = Object.keys(shards);
+    const fileName = manifest.fileName;
+
+    const shardAuditData = shaIds.reduce((acc, shaId) => {
+      acc[shaId] = {};
+
+      shards[shaId].forEach((shardId) => {
+        acc[shaId][shardId] = false;
+      });
+
+      return acc;
+    }, {});
+
+    this.auditShardsGroup(shards, 0, shardAuditData);
+  }
+  /**
+   * Tests the redudant copies of the original shard for data integrity.
+   * @param {shards} Object - Shard content SHA keys with
+   * array of redundant shard ids
+   * @param {shaIdx} Number - Index of the current
+   * @param {shardAuditData} Object - same as shards param except instead of an
+   * array of shard ids it's an object of shard ids and their audit status
+  */
+  auditShardsGroup(shards, shaIdx, shardAuditData) {
+    let shardDupIdx = 0;
+    const shaId = Object.keys(shards)[shaIdx];
+
+    this.auditShard(shards, shardDupIdx, shaId, shaIdx, shardAuditData);
+  }
+
+  auditShard(shards, shardDupIdx, shaId, shaIdx, shardAuditData) {
+    const shardId = shards[shaId][shardDupIdx];
+
+    this.kadenceNode.iterativeFindValue(shardId, (err, value, responder) => {
+      let kadNodeTarget = value.value;
+      this.kadenceNode.getOtherBatNodeContact(kadNodeTarget, (err, batNode) => {
+        this.auditShardData(batNode, shards, shaIdx, shardDupIdx, shardAuditData)
+      })
+    })
+  }
+
+  auditShardData(targetBatNode, shards, shaIdx, shardDupIdx, shardAuditData) {
+    let client = this.connect(targetBatNode.port, targetBatNode.host);
+
+    const shaKeys = Object.keys(shards);
+    const shaId = shaKeys[shaIdx];
+    const shardId = shards[shaId][shardDupIdx]; // id of a redundant shard for shaId
+    const moreShaGroups = shaKeys.length > shaIdx + 1;
+    const moreShardsInGroup = shards[shaId].length > shardDupIdx + 1;
+
+    let message = {
+      messageType: "AUDIT_FILE",
+      fileName: shardId
+    };
+
+    client.write(JSON.stringify(message), (err) => {
+      if (err) { throw err; }
+    })
+
+    client.on('data', (data) => {
+      const hostShardSha1 = data.toString('utf8');
+      // Check that shard content matches original content SHA
+      if (hostShardSha1 === shaId) {
+        shardAuditData[shaId][shardId] = true;
+      }
+
+      // Continuing audit logic
+      if (moreShardsInGroup) {
+        this.auditShard(shards, shardDupIdx + 1, shaId, shaIdx, shardAuditData);
+      } else {
+        if (moreShaGroups) {
+          this.auditShardsGroup(shards, shaIdx + 1, shardAuditData);
+        } else {
+          const dataValid = this.auditResults(shardAuditData, shaKeys);
+          if (dataValid) {
+            console.log('Passed audit!');
+          } else {
+            console.log('Failed Audit');
+          }
+        }
+      }
+    })
+  }
```

Comment:
I agree that the `auditShardData` contains a lot of logic and it will be good if we can move part of the logic elsewhere(like the example in your comment)

Reply:
Yeah, it took a little while but I agreed and I did that. See the latest commit.